### PR TITLE
[FIX] mail: Handle bounces from Gmail

### PR DESCRIPTION
**Pull request by yajo**
Gmail bounces include the `Message-Id` header in a section typed `text/rfc822-headers` instead of a `message/rfc822`. Some other email providers do this too, as it is a standard.

According to https://quimby.gnus.org/rfc/rfc1892.txt that section's body actually contains the headers from the bounced email.

Without this patch, bounces from Gmail cannot find the `bounced_message_id`, and thus the little email envelope icon doesn't turn red, nor do the email resend features trigger in the UI. You could be sending an invoice and never knowing it was bounced.

After this patch, bounces from mail servers that return the `text/rfc822-headers` part will be handled properly.

@Tecnativa TT21170 @Tardo @rafaelbn

--
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

**Comment by pedrobaeza** [comment]
@tde-banana-odoo can you take a look to this, please?

**Comment by nim-odoo** [comment]
I am not able to reproduce the issue. I am using our SMTP/IMAP server and send a message to a non-existing gmail address. This is logged as expected, and the envelope turns red:

```
2020-01-20 11:51:35,096 15695 INFO 11ent odoo.addons.mail.models.mail_thread: Routing mail from MAILER-DAEMON@mail2.odoo.com (Mail Delivery System) to nim+33-account.invoice-11@odoo.com with Message-Id <20200120115037.6B71CA43F2@mail2.odoo.com>: bounced mail from mail 33, model: account.invoice, thread_id: 11: dest sergsedrgdwrgwdghxfthdwrgdwerg@gmail.com (partner res.partner(53,))
```

The `content_type` is `message/rfc822`, as expected.

Do you have any email address I could test the bounce on?

**Comment by rafaelbn** [comment]
@Yajo can you answer here? is this also happening in v12?

**Comment by Yajo** [comment]
Sorry I asked our customer for some test credentials in their mail server and he never answered, give me a little more time to see if I can get that.

**Comment by Yajo** [comment]
Sorry for the long silence.

The problem is not strictly related to gmail as it seems. To reproduce it:

1. Make sure your postfix server is >= v2.5
1. Configure it with `bounce_size_limit=1`. This will remove the original mail from any bounce from your mailer daemon and it will include just the headers (producing a `text/rfc822-headers` section).
1. Send an email to any non-existing address (not just gmail).

You'll see this section in the mail bounce and all the symptoms described in the original message (it had no special relation with gmail after all, sorry for confusion).

Yes, v12+ is affected too.

**Comment by pedrobaeza** [comment]
@nim-odoo any news about this?

**Comment by yelizariev** [comment]
@Yajo do you have a raw sample for such bounced email? I'd like to have a test for this, similar to message/rfc822, 

https://github.com/odoo/odoo/blob/889e7c76b2fb2cb9523b7e74961356bca50be417/addons/mail/tests/test_mail_gateway.py#L758-L762

opw-2344252

**Comment by Yajo** [comment]
I do have one, but I cannot publish it because it contains private info. Let me see if I can anonymize it.

**Comment by Yajo** [comment]
OK I managed to do it:

<details>

```eml
Date: Tue, 24 Dec 2019 11:32:07 +0100 (CET)
MIME-Version: 1.0
Content-Type: multipart/mixed; boundary=16063919151.b32bE0eD.7
Content-Transfer-Encoding: 7bit
Subject: Undelivered Mail Returned to Sender
From: MAILER-DAEMON@example.com (Mail Delivery System)
To: bounce+82240-account.invoice-19177@mycompany.example.com
Message-Id: <20191224103207.415713014C@example.com>
Return-Path: <MAILER-DAEMON>
Delivered-To: odoo+82240-account.invoice-19177@mycompany.example.com
Received: by example.com (Postfix) id 415713014C; Tue, 24 Dec
 2019 11:32:07 +0100 (CET)
Auto-Submitted: auto-replied


--16063919151.b32bE0eD.7
MIME-Version: 1.0
Content-Type: multipart/alternative; boundary=16063919150.2cD3F37.7
Content-Transfer-Encoding: 7bit
Content-ID: <16063919152.fD96.7@8f286b7b7880>


--16063919150.2cD3F37.7
Content-Type: text/plain; charset=US-ASCII
Content-Disposition: inline
Content-Transfer-Encoding: 8bit

This is the mail system at host example.com.

I'm sorry to have to inform you that your message could not
be delivered to one or more recipients. It's attached below.

For further assistance, please send mail to postmaster.

If you do so, please include this problem report. You can
delete your own text from the attached returned message.

                   The mail system

<pepitogrillo9876@tecnativa.com>: host aspmx.l.google.com[173.194.76.27] said:
    550-5.1.1 The email account that you tried to reach does not exist. Please
    try 550-5.1.1 double-checking the recipient's email address for typos or
    550-5.1.1 unnecessary spaces. Learn more at 550 5.1.1
    https://support.google.com/mail/?p=NoSuchUser a6si19853016wrt.319 - gsmtp
    (in reply to RCPT TO command)

--16063919150.2cD3F37.7
Content-Type: application/rtf
Content-Disposition: inline
Content-Transfer-Encoding: base64

e1xydGYxXGFuc2lcYW5zaWNwZzEyNTJcZnJvbXRleHQgXGZiaWRpcyBcZGVmZjB7XGZvbnR0YmwK
DXtcZjBcZnN3aXNzIEFyaWFsO30KDXtcZjFcZm1vZGVybiBDb3VyaWVyIE5ldzt9Cg17XGYyXGZu
aWxcZmNoYXJzZXQyIFN5bWJvbDt9Cg17XGYzXGZtb2Rlcm5cZmNoYXJzZXQwIENvdXJpZXIgTmV3
O319Cg17XGNvbG9ydGJsXHJlZDBcZ3JlZW4wXGJsdWUwO1xyZWQwXGdyZWVuMFxibHVlMjU1O30K
DVx1YzFccGFyZFxwbGFpblxkZWZ0YWIzNjAgXGYwXGZzMjAgVGhpcyBpcyB0aGUgbWFpbCBzeXN0
ZW0gYXQgaG9zdCBleGFtcGxlLmNvbS5ccGFyCg1ccGFyCg1JJ20gc29ycnkgdG8gaGF2ZSB0byBp
bmZvcm0geW91IHRoYXQgeW91ciBtZXNzYWdlIGNvdWxkIG5vdFxwYXIKDWJlIGRlbGl2ZXJlZCB0
byBvbmUgb3IgbW9yZSByZWNpcGllbnRzLiBJdCdzIGF0dGFjaGVkIGJlbG93LlxwYXIKDVxwYXIK
DUZvciBmdXJ0aGVyIGFzc2lzdGFuY2UsIHBsZWFzZSBzZW5kIG1haWwgdG8gcG9zdG1hc3Rlci5c
cGFyCg1ccGFyCg1JZiB5b3UgZG8gc28sIHBsZWFzZSBpbmNsdWRlIHRoaXMgcHJvYmxlbSByZXBv
cnQuIFlvdSBjYW5ccGFyCg1kZWxldGUgeW91ciBvd24gdGV4dCBmcm9tIHRoZSBhdHRhY2hlZCBy
ZXR1cm5lZCBtZXNzYWdlLlxwYXIKDVxwYXIKDSAgICAgICAgICAgICAgICAgICBUaGUgbWFpbCBz
eXN0ZW1ccGFyCg1ccGFyCg08cGVwaXRvZ3JpbGxvOTg3NkB0ZWNuYXRpdmEuY29tPjogaG9zdCBh
c3BteC5sLmdvb2dsZS5jb21bMTczLjE5NC43Ni4yN10gc2FpZDpccGFyCg0gICAgNTUwLTUuMS4x
IFRoZSBlbWFpbCBhY2NvdW50IHRoYXQgeW91IHRyaWVkIHRvIHJlYWNoIGRvZXMgbm90IGV4aXN0
LiBQbGVhc2VccGFyCg0gICAgdHJ5IDU1MC01LjEuMSBkb3VibGUtY2hlY2tpbmcgdGhlIHJlY2lw
aWVudCdzIGVtYWlsIGFkZHJlc3MgZm9yIHR5cG9zIG9yXHBhcgoNICAgIDU1MC01LjEuMSB1bm5l
Y2Vzc2FyeSBzcGFjZXMuIExlYXJuIG1vcmUgYXQgNTUwIDUuMS4xXHBhcgoNICAgIGh0dHBzOi8v
c3VwcG9ydC5nb29nbGUuY29tL21haWwvP3A9Tm9TdWNoVXNlciBhNnNpMTk4NTMwMTZ3cnQuMzE5
IC0gZ3NtdHBccGFyCg0gICAgKGluIHJlcGx5IHRvIFJDUFQgVE8gY29tbWFuZClccGFyCg19

--16063919150.2cD3F37.7--

--16063919151.b32bE0eD.7
Content-ID: <16063919153.DbAa.7@8f286b7b7880>
Content-Type: message/delivery-status; name=details.txt
Content-Disposition: attachment; filename=details.txt
Content-Transfer-Encoding: base64

UmVwb3J0aW5nLU1UQTogZG5zOyBleGFtcGxlLmNvbQ0KWC1Qb3N0Zml4LVF1ZXVlLUlEOiBCOUMz
NTMwNTkyDQpYLVBvc3RmaXgtU2VuZGVyOiByZmM4MjI7IGJvdW5jZSs4MjI0MC1hY2NvdW50Lmlu
dm9pY2UtMTkxNzdAbXljb21wYW55LmV4YW1wbGUuY29tDQpBcnJpdmFsLURhdGU6IFR1ZSwgMjQg
RGVjIDIwMTkgMTE6MzI6MDYgKzAxMDAgKENFVCkNCg0KRmluYWwtUmVjaXBpZW50OiByZmM4MjI7
IHBlcGl0b2dyaWxsbzk4NzZAdGVjbmF0aXZhLmNvbQ0KT3JpZ2luYWwtUmVjaXBpZW50OiByZmM4
MjI7cGVwaXRvZ3JpbGxvOTg3NkB0ZWNuYXRpdmEuY29tDQpBY3Rpb246IGZhaWxlZA0KU3RhdHVz
OiA1LjEuMQ0KUmVtb3RlLU1UQTogZG5zOyBhc3BteC5sLmdvb2dsZS5jb20NCkRpYWdub3N0aWMt
Q29kZTogc210cDsgNTUwLTUuMS4xIFRoZSBlbWFpbCBhY2NvdW50IHRoYXQgeW91IHRyaWVkIHRv
IHJlYWNoIGRvZXMNCiAgICBub3QgZXhpc3QuIFBsZWFzZSB0cnkgNTUwLTUuMS4xIGRvdWJsZS1j
aGVja2luZyB0aGUgcmVjaXBpZW50J3MgZW1haWwNCiAgICBhZGRyZXNzIGZvciB0eXBvcyBvciA1
NTAtNS4xLjEgdW5uZWNlc3Nhcnkgc3BhY2VzLiBMZWFybiBtb3JlIGF0IDU1MCA1LjEuMQ0KICAg
IGh0dHBzOi8vc3VwcG9ydC5nb29nbGUuY29tL21haWwvP3A9Tm9TdWNoVXNlciBhNnNpMTk4NTMw
MTZ3cnQuMzE5IC0gZ3NtdHANCg==

--16063919151.b32bE0eD.7
Content-ID:
Content-Type: text/rfc822-headers; name="Undelivered Message Headers.txt"
Content-Disposition: attachment; filename="Undelivered Message Headers.txt"
Content-Transfer-Encoding: base64

UmV0dXJuLVBhdGg6IDxib3VuY2UrODIyNDAtYWNjb3VudC5pbnZvaWNlLTE5MTc3QG15Y29tcGFu
eS5leGFtcGxlLmNvbT4NClJlY2VpdmVkOiBmcm9tIGxvY2FsaG9zdCAobG9jYWxob3N0IFsxMjcu
MC4wLjFdKQ0KCWJ5IGV4YW1wbGUuY29tIChQb3N0Zml4KSB3aXRoIEVTTVRQIGlkIEI5QzM1MzA1
OTINCglmb3IgPHBlcGl0b2dyaWxsbzk4NzZAdGVjbmF0aXZhLmNvbT47IFR1ZSwgMjQgRGVjIDIw
MTkgMTE6MzI6MDYgKzAxMDAgKENFVCkNClgtVmlydXMtU2Nhbm5lZDogRGViaWFuIGFtYXZpc2Qt
bmV3IGF0IGV4YW1wbGUuY29tDQpSZWNlaXZlZDogZnJvbSBleGFtcGxlLmNvbSAoWzEyNy4wLjAu
MV0pDQoJYnkgbG9jYWxob3N0IChleGFtcGxlLmNvbSBbMTI3LjAuMC4xXSkgKGFtYXZpc2QtbmV3
LCBwb3J0IDEwMDI2KQ0KCXdpdGggRVNNVFAgaWQgNFpubXowUTl1eHlUIGZvciA8cGVwaXRvZ3Jp
bGxvOTg3NkB0ZWNuYXRpdmEuY29tPjsNCglUdWUsIDI0IERlYyAyMDE5IDExOjMyOjA2ICswMTAw
IChDRVQpDQpSZWNlaXZlZDogZnJvbSBsb2NhbGhvc3QubG9jYWxkb21haW4gKHN0YXRpYy4xNjcu
MTQwLjkuNS5jbGllbnRzLnlvdXItc2VydmVyLmRlIFs1LjkuMTQwLjE2N10pDQoJKEF1dGhlbnRp
Y2F0ZWQgc2VuZGVyOiBvZG9vQG15Y29tcGFueS5leGFtcGxlLmNvbSkNCglieSBleGFtcGxlLmNv
bSAoUG9zdGZpeCkgd2l0aCBFU01UUFNBIGlkIDM5RTYyMzAxMzUNCglmb3IgPHBlcGl0b2dyaWxs
bzk4NzZAdGVjbmF0aXZhLmNvbT47IFR1ZSwgMjQgRGVjIDIwMTkgMTE6MzI6MDYgKzAxMDAgKENF
VCkNClJlY2VpdmVkOiBmcm9tIG15Y29tcGFueS5leGFtcGxlLmNvbSAoc3BhY29ycC1vZG9vLTEy
X29kb29fMS5zcGFjb3JwLW9kb28tMTJfZGVmYXVsdCBbMTcyLjI2LjAuNV0pDQoJYnkgbG9jYWxo
b3N0IChQb3N0Zml4KSB3aXRoIEVTTVRQIGlkIEVFMjI2MTUyMThBDQoJZm9yIDxwZXBpdG9ncmls
bG85ODc2QHRlY25hdGl2YS5jb20+OyBUdWUsIDI0IERlYyAyMDE5IDEwOjMyOjA1ICswMDAwIChV
VEMpDQpDb250ZW50LVR5cGU6IG11bHRpcGFydC9taXhlZDsgYm91bmRhcnk9Ij09PT09PT09PT09
PT09PTM2MDA3NTkyMjYxNTg1NTE5OTQ9PSINCk1JTUUtVmVyc2lvbjogMS4wDQpNZXNzYWdlLUlk
OiA8ODYxODc4MTc1ODIzMTQ4LjE1NzcxODM1MjUuNzM2MDA1NzgzMDgxMDU1LW9wZW5lcnAtMTkx
NzctYWNjb3VudC5pbnZvaWNlQG15Y29tcGFueS5leGFtcGxlLmNvbT4NCnJlZmVyZW5jZXM6IDwx
NTcxODE0NDgxLjE4OTI4MTk0MDQ2MDIwNS43OTk1ODI0NDEyMzg0Njctb3BlbmVycC0xOTE3Ny1h
Y2NvdW50Lmludm9pY2VAcGVyc3VtLmVzPg0KU3ViamVjdDogUFJVRUJBIDMgTVkgQ09NUEFOWSBF
VVJPUEUgUy5MLiBGYWN0dXJhIChSZWYgRTEvMTkzOTI2KQ0KRnJvbTogIlNpbiBBZ2VudGUgQXNp
Z25hZG8iIDxub3JlcGx5K3NyZ2x2cnotZ21haWwuY29tQG15Y29tcGFueS5leGFtcGxlLmNvbT4N
ClJlcGx5LVRvOiAiTVkgQ09NUEFOWSBFVVJPUEUgUy5MLiBFMS8xOTM5MjYgIiA8c3BhY29ycEBt
eWNvbXBhbnkuZXhhbXBsZS5jb20+DQpUbzogIkNMSUVOVEVTIFZBUklPUyIgPHBlcGl0b2dyaWxs
bzk4NzZAdGVjbmF0aXZhLmNvbT4NCkRhdGU6IFR1ZSwgMjQgRGVjIDIwMTkgMTA6MzI6MDUgLTAw
MDANClgtT2Rvby1PYmplY3RzOiBhY2NvdW50Lmludm9pY2UtMTkxNzcNCg==

--16063919151.b32bE0eD.7--
```

</details>

**Comment by Yajo** [comment]
In which version(s) did the fix land?

**Comment by yelizariev** [comment]
13.0+

**Comment by yelizariev** [comment]
For v12 this PR could be applied without modifications https://github.com/odoo/odoo/pull/62551/files, but I didn't because v12 is not structured well to make a proper test. If we get a support request for v12, it would be landed to v12 too.

And v11 is out of maintenance as you know

File: addons/mail/models/mail_thread.py [hunk 1]
```diff
@@ -1082,7 +1083,11 @@ def message_route(self, message, message_dict, model=None, thread_id=None, custo
 
                 mail_message = self.env['mail.message']
                 if email_part:
-                    email = email_part.get_payload()[0]
+                    if email_part.get_content_type() == 'text/rfc822-headers':
+                        # Convert the message body into a message itself
+                        email = message_from_string(email_part.get_payload())
```
Comment (Julien00859) [inline addons/mail/models/mail_thread.py:1088]: Should had been `email_part.get_content()`

Comment (Julien00859) [reply addons/mail/models/mail_thread.py:1088]: also no unittest...

**Pull request closed without merging**